sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,312 @@
+
+        Parameters
+        ----------
+
+        X : array-like, shape (n_samples, n_features), optional
+            The query sample or samples to compute the Local Outlier Factor
+            w.r.t. the training samples. If None, makes prediction on the
+            training data without considering them as their own neighbors.
+
+        Returns
+        -------
+        lof_scores : array of shape (n_samples,)
+            The Local Outlier Factor of each input samples. The lower,
+            the more abnormal.
+        """
+        check_is_fitted(self, ["threshold_", "decision_function_train_",
+                               "_k_distance_value_fit_X_",
+                               "neighbors_indices_fit_X_" ])
+
+        # if X is None, make prediction on X=X_train without considered each
+        # query point as its own neighbor.
+        if X is None:
+            return self.decision_function_train_
+
+        else:
+            X = check_array(X, accept_sparse='csr')
+            return -self._local_outlier_factor(X)  # as bigger is better
+
+    def _k_distance(self, X=None):
+        """
+        Compute the k_distance and the neighborhood of query samples wrt
+        training samples self._fit_X.
+        If X=None, neighbors of each sample self._fit_X are returned.
+        In this case, the query point is not considered its own neighbor.
+        """
+        distances, neighbors_indices = self.kneighbors(
+            X=X, n_neighbors=self.n_neighbors)
+        k_dist = distances[:, self.n_neighbors - 1]
+
+        return k_dist, neighbors_indices
+
+    def _local_reachability_density(self, X, neighbors_indices):
+        """The local reachability density (LRD)
+
+        The LRD of a sample is the inverse of the average reachability
+        distance of its k-nearest neighbors.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The samples to compute the LRD w.r.t. self._fit_X
+            If None, compute the LRD of self._fit_X w.r.t. itself.
+            (In this case samples are not considered in their own
+            neighborhood)
+
+        neighbors_indices : array-like of shape (n_samples, self.n_neighbors)
+            neighbors indices of X (or of self._fit_X if X is None)
+            Added as a parameter to avoid re-computing it with 
+            self._k_distance(X)[1] when X is not None (if X is None, it is
+            equal to self.neighbors_indices_fit_X_)
+
+        Returns
+        -------
+        local_reachability_density : array, shape (n_samples,)
+            The local reachability density of each sample.
+        """
+        if X is None:
+            X = self._fit_X
+
+        n_samples = X.shape[0]
+        reach_dist_array = np.zeros((n_samples, self.n_neighbors))
+
+        for j in range(n_samples):
+            neighbors_number = -1
+            dist_j = pairwise_distances(X[j,:],
```

Comment:
This still remains a problem, it causes depreciation warnings. Is there a way to do `X[j, :].reshape((1,-1))` but in a way it can handle sparse matrices? ping @raghavrv